language-plutus-core/src/Language/PlutusCore/Quote.hs

```diff
@@ -0,0 +1,49 @@
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+
+module Language.PlutusCore.Quote (
+              runQT
+            , runQ
+            , freshName
+            , freshTyName
+            , parse
+            , QT
+            , Q
+            ) where
+
+import           Control.Monad.Except
+import           Control.Monad.State
+import qualified Data.ByteString.Lazy       as BSL
+import           Language.PlutusCore.Lexer  (AlexPosn)
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Parser (ParseError, parseST)
+import           Language.PlutusCore.Type
+import           Data.Functor.Identity
+import           PlutusPrelude
+
+-- | The "quotation" monad transformer. This allows creation of fresh names and parsing.
+newtype QT m a = QT { unQT :: StateT IdentifierState m a }
+  deriving (Functor, Applicative, Monad, MonadTrans, MonadState IdentifierState)
+
+-- | Run a quote from an empty identifier state.
+runQT ::  (Monad m) => QT m a -> m a
+runQT q = evalStateT (unQT q) emptyIdentifierState
+
+type Q a = QT Identity a
```

Comment:
I was deliberately punning on the TH `Q`, since this is our version of it, but I don't feel strongly about it. I'll rename it to `Quote`. I would like to keep the quoting association, since this is also the monad that people will use when using the quoter.